sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -695,6 +704,15 @@ int64_t CurlSession::Read(Context const& context, uint8_t* buffer, int64_t count
 int64_t CurlSession::ReadFromSocket(Context const& context, uint8_t* buffer, int64_t bufferSize)
 {
   // loop until read result is not CURLE_AGAIN
+  // Next loop is expected to be called at most 2 times:
+  // The first time it calls `curl_easy_recv()`, if it returns CURLE_AGAIN it would call
+  // `pollSocketUntilEventOrTimeout` and wait for socket to be ready to read.
+  // `pollSocketUntilEventOrTimeout` will then handle cancelation token.
+  // If socket is not ready before the timeout, Exception is thrown.
+  // When socket is ready, it calls curl_easy_recv() again (second loop iteration). It is not
+  // expected to return CURLE_AGAIN (since socket is ready), so, a chuck of data will be downloaded
```

Comment:
> When socket is ready, it calls curl_easy_recv() again (second loop iteration). It is not expected to return CURLE_AGAIN (since socket is ready)

How sure are we that this is true?

Reply:
if for any weird reason it returns CURLE_AGAIN, we would just wait for socket again.
However, I feel 100% sure this won't cause, at any case, something like an infinite loop where it will return CURLE_AGAIN  forever and then waitForSocket keeps returning ready.

Reply:
I understand we may not have an infinite loop,  but if it starts iterating a lot, then we may not be respecting the context cancellation. That's why I was thinking we should be paranoid and move it into the inner loop, but as long as you are confident it won't happen, than leaving it in the outer-loop is fine.